### DISCUSSION TITLE
Modelling ADTs: union types vs mixin

### DISCUSSION THREAD

**Discussion by ShahOdin** [disneystreaming/smithy4s#586] [General]
I want to end up generating the following scala code:

```scala
saeld trait Foo {
  val internal_id: String
  val external_id: String 
}

object Foo{
  case object Bar extends Foo {
    val internal_id: String = "1"
    val external_id: String = "2"
  }

  case object Baz extends Foo {
    val internal_id: String = "10"
    val external_id: String = "20"
  }
}
```

as closely as possible. What would be the cleanest/preferred way of achieving this in smithy? Think both `union types` and mixins would give me something close enough 🤔 

ideally, It'd be nice if we can have an enumeratum style `Foo.all` method generated for us as well. that would return `Seq(Bar, Baz)`

**Comment by lewisjkl** [comment]
Hey ShahOdin, you aren't going to be able to generate something exactly like this. The primary reason is that what you want isn't even expressible in smithy. For example,

```smithy
structure Bar {
  internal_id: String = "1"
}
```

this structure does not represent an *instance* of a class, but rather a definition of a class. This means that the `"1"` here represents a default parameter as seen in:

```scala
final case class Bar(internal_id: String = "1")
```

So you definitely can get close to what you want with mixins (though mixins do not generate sealed traits, but rather open traits) and unions, but not exactly.

I hope this helps, but feel free to follow up with any questions!

**Reply by kubukoz** [reply]
Wouldn't this be possible with `@adtMember` and mixins?

**Reply by kubukoz** [reply]
ah, no, I misunderstood the thing. Forget it ;)
